### PR TITLE
feat: dynamic rebalancing of clusters across shards

### PR DESCRIPTION
**Pull request by ishitasequeira**
This PR is an implementation of the following proposal:
https://github.com/argoproj/argo-cd/blob/master/docs/proposals/rebalancing-clusters-across-shards-dynamically.md

*Note*: As implemented in this PR, the new dynamic rebalancing feature is enabled by migrating the Argo CD Application controller from StatefulSet to a Deployment. Migration instructions are provided as part of the docs. This is an `alpha` feature and in the future, we may choose to return to using a StatefulSet instead.

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

File: cmd/argocd-application-controller/commands/argocd_application_controller.go [hunk 1]
```diff
@@ -206,20 +209,50 @@ func NewCommand() *cobra.Command {
 }
 
 func getClusterFilter(kubeClient *kubernetes.Clientset, settingsMgr *settings.SettingsManager, shardingAlgorithm string) sharding.ClusterFilterFunction {
-	replicas := env.ParseNumFromEnv(common.EnvControllerReplicas, 0, 0, math.MaxInt32)
+
+	var replicas int
 	shard := env.ParseNumFromEnv(common.EnvControllerShard, -1, -math.MaxInt32, math.MaxInt32)
+
+	appControllerDeployment, _ := kubeClient.AppsV1().Deployments(settingsMgr.GetNamespace()).Get(context.Background(), common.ApplicationController, metav1.GetOptions{})
```
Comment (leoluz) [inline cmd/argocd-application-controller/commands/argocd_application_controller.go:216]: This seems a bit fragile as if someone changes the application controller name, it would break the rebalancing feature. Maybe we shouldn't base on the name but on a predefined label? Or maybe something else to make it more robust? WDYT?

Comment (ishitasequeira) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go:216]: A very valid point. The same is true for a predefined label value. As discussed over slack, I will add an e2e test to make sure we validate on the controller name or fail the test, thus failing the build.

File: controller/sharding/sharding.go [hunk 2]
```diff
@@ -162,3 +186,156 @@ func createClusterIndexByClusterIdMap(db db.ArgoDB) map[string]int {
...
+/*
+ * GetShardFromConfigMap finds the shard number from the shard mapping configmap. If the shard mapping configmap does not exist,
+ * the function creates the shard mapping configmap. If the shard number is set as an environment variable for the application controller pod,
+ * we use the shard number to update the mapping.
+ */
+func GetShardFromConfigMap(kubeClient *kubernetes.Clientset, settingsMgr *settings.SettingsManager, replicas, shard int) (int, error) {
+
+	hostname, err := osHostnameFunction()
+	if err != nil {
+		return -1, err
```
Comment (leoluz) [inline controller/sharding/sharding.go:199]: Please add context to returned errors

File: controller/sharding/sharding.go [hunk 3]
```diff
@@ -162,3 +186,156 @@ func createClusterIndexByClusterIdMap(db db.ArgoDB) map[string]int {
...
+	}
+
+	// fetch the shard mapping configMap
+	shardMappingCM, err := kubeClient.CoreV1().ConfigMaps(settingsMgr.GetNamespace()).Get(context.Background(), common.ArgoCDAppControllerShardConfigMapName, metav1.GetOptions{})
+
+	if err != nil {
+		log.Infof("shard mapping configmap %s not found. Creating default shard mapping configmap.", common.ArgoCDAppControllerShardConfigMapName)
+
+		// if the shard is not set as an environment variable, set the default value of shard to 0 for generating default CM
+		if shard == -1 {
```
Comment (leoluz) [inline controller/sharding/sharding.go:209]: What shard == -1 means exactly? I think that this is an importat info to add in the godoc of this function.

File: controller/sharding/sharding.go [hunk 4]
```diff
@@ -162,3 +186,156 @@ func createClusterIndexByClusterIdMap(db db.ArgoDB) map[string]int {
...
+		if err != nil {
+			return -1, fmt.Errorf("error marshalling data of shard mapping ConfigMap: %s", err)
+		}
+		shardMappingCM.Data[ShardControllerMappingKey] = string(updatedShardMappingData)
+
+		_, err = kubeClient.CoreV1().ConfigMaps(settingsMgr.GetNamespace()).Update(context.Background(), shardMappingCM, metav1.UpdateOptions{})
+		if err != nil {
+			return -1, err
+		}
+		return shard, err
```
Comment (leoluz) [inline controller/sharding/sharding.go:241]: Please add context to returned errors

**Review by leoluz — CHANGES_REQUESTED** [review]
Thank you Ishita! This is going to be an awesome feature!
Apart from my comments there are few things pending in this PR:
- Documentation
- Tests (lots of them) :)

File: cmd/argocd-application-controller/commands/argocd_application_controller.go [hunk 5]
```diff
@@ -206,20 +209,49 @@ func NewCommand() *cobra.Command {
...
 func getClusterFilter(kubeClient *kubernetes.Clientset, settingsMgr *settings.SettingsManager, shardingAlgorithm string) sharding.ClusterFilterFunction {
-	replicas := env.ParseNumFromEnv(common.EnvControllerReplicas, 0, 0, math.MaxInt32)
+
+	var replicas int
 	shard := env.ParseNumFromEnv(common.EnvControllerShard, -1, -math.MaxInt32, math.MaxInt32)
+
+	applicationControllerName := env.StringFromEnv(common.EnvAppControllerName, common.DefaultApplicationControllerName)
+	appControllerDeployment, _ := kubeClient.AppsV1().Deployments(settingsMgr.GetNamespace()).Get(context.Background(), applicationControllerName, metav1.GetOptions{})
+
+	if appControllerDeployment != nil {
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go:219]: We try to fetch the application deployment. The idea here is to try and fetch the application controller deployment. If we do not find the deployment, we assume the controller to be running as part of StatefulSet.

File: cmd/argocd-application-controller/commands/argocd_application_controller.go [hunk 6]
```diff
@@ -206,20 +209,49 @@ func NewCommand() *cobra.Command {
 }
 
 func getClusterFilter(kubeClient *kubernetes.Clientset, settingsMgr *settings.SettingsManager, shardingAlgorithm string) sharding.ClusterFilterFunction {
-	replicas := env.ParseNumFromEnv(common.EnvControllerReplicas, 0, 0, math.MaxInt32)
+
+	var replicas int
 	shard := env.ParseNumFromEnv(common.EnvControllerShard, -1, -math.MaxInt32, math.MaxInt32)
+
+	applicationControllerName := env.StringFromEnv(common.EnvAppControllerName, common.DefaultApplicationControllerName)
+	appControllerDeployment, _ := kubeClient.AppsV1().Deployments(settingsMgr.GetNamespace()).Get(context.Background(), applicationControllerName, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go:217]: Yes, we are trying to make sure we do not break the controller if someone decides to use StatefulSet (trying to be backwards compatible). We just try to fetch the application deployment. The idea here is to try and fetch the application controller deployment. If we do not fetch the deployment, we assume the controller to be running as part of StatefulSet. 

It does makes sense to check for the error message (add a check on Not Found error) and add debug logs on the approach being used. I will add them.

File: cmd/argocd-application-controller/commands/argocd_application_controller.go [hunk 7]
```diff
@@ -206,20 +209,49 @@ func NewCommand() *cobra.Command {
...
+
+	if appControllerDeployment != nil {
+		replicas = int(*appControllerDeployment.Spec.Replicas)
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go:220]: I didn't think of any edge case where the `appControllerDeployment.Spec.Replicas` could be nil, but it's a good idea to add. I will add this check.

File: cmd/argocd-application-controller/commands/argocd_application_controller.go [hunk 8]
```diff
@@ -206,20 +209,49 @@ func NewCommand() *cobra.Command {
...
+		replicas = env.ParseNumFromEnv(common.EnvControllerReplicas, 0, 0, math.MaxInt32)
+	}
+
 	var clusterFilter func(cluster *v1alpha1.Cluster) bool
 	if replicas > 1 {
-		if shard < 0 {
+		// check for shard mapping using configmap if application-controller is a deployment
+		// else use existing logic to infer shard from pod name if application-controller is a statefulset
+		if appControllerDeployment != nil {
+			retryCount := 3
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go:230]: I was thinking of making a global variable for this but not configurable. This `retryCount` represents the number of times we retry in every heartbeat. IMO, it would retry for `retryCount` times in the next heartbeat of Readiness probe check even if it fails 3 times here. WDYT?

File: cmd/main.go [hunk 9]
```diff
@@ -34,7 +35,7 @@ func main() {
 		command = cli.NewCommand()
 	case "argocd-server":
 		command = apiserver.NewCommand()
-	case "argocd-application-controller":
+	case common.ApplicationController:
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply cmd/main.go:38]: I will probably change all cases to constant.

File: controller/appcontroller.go [hunk 10]
```diff
@@ -201,11 +202,25 @@ func NewApplicationController(
 			}
 		},
 	})
+
+	readinessHealthCheck := func(r *http.Request) error {
+		applicationControllerName := env.StringFromEnv(common.EnvAppControllerName, common.DefaultApplicationControllerName)
+		appControllerDeployment, _ := kubeClientset.AppsV1().Deployments(settingsMgr.GetNamespace()).Get(context.Background(), applicationControllerName, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply controller/appcontroller.go:208]: We currently do not have a deployment Informer configured. I can try configuring one.

File: controller/appcontroller.go [hunk 11]
```diff
@@ -201,11 +202,25 @@ func NewApplicationController(
...
+		if appControllerDeployment != nil {
+			if appControllerDeployment.Spec.Replicas != nil && int(*appControllerDeployment.Spec.Replicas) <= 0 {
+				return fmt.Errorf("application controller deployment replicas is not set or is less than 0, replicas: %d", appControllerDeployment.Spec.Replicas)
+			}
+			shard := env.ParseNumFromEnv(common.EnvControllerShard, -1, -math.MaxInt32, math.MaxInt32)
+			if _, err := sharding.GetShardFromConfigMap(kubeClientset.(*kubernetes.Clientset), settingsMgr, int(*appControllerDeployment.Spec.Replicas), shard); err != nil {
+				return err
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply controller/appcontroller.go:215]: Most of the returned messages from `GetShardFromConfigMap` are returning the errors with context. I will update the missing errors inside `GetShardFromConfigMap` with context. That's the reason I did not add a context to this error message.

File: controller/appcontroller.go [hunk 12]
```diff
@@ -201,11 +202,25 @@ func NewApplicationController(
...
+		}
+		return nil
+	}
+
 	metricsAddr := fmt.Sprintf("0.0.0.0:%d", metricsPort)
 	var err error
-	ctrl.metricsServer, err = metrics.NewMetricsServer(metricsAddr, appLister, ctrl.canProcessApp, func(r *http.Request) error {
-		return nil
-	}, metricsApplicationLabels)
+	ctrl.metricsServer, err = metrics.NewMetricsServer(metricsAddr, appLister, ctrl.canProcessApp, readinessHealthCheck, metricsApplicationLabels)
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply controller/appcontroller.go:223]: So, this readiness check is basically used to update the readiness probe of the application-controller pod. Initially we were just returning `nil` which marked the pod as `Ready` always. But now, we mark the controller pod as `Ready` only if the controller is able to claim a shard or update the heartbeat timestamp in the ConfigMap. Also, we perform the readiness check only if the application-controller is a Deployment and not StatefulSet.

File: controller/sharding/sharding.go [hunk 13]
```diff
@@ -162,3 +186,156 @@ func createClusterIndexByClusterIdMap(db db.ArgoDB) map[string]int {
 	}
 	return clusterIndexedByClusterId
 }
+
+/*
+ * GetShardFromConfigMap finds the shard number from the shard mapping configmap. If the shard mapping configmap does not exist,
+ * the function creates the shard mapping configmap. If the shard number is set as an environment variable for the application controller pod,
+ * we use the shard number to update the mapping.
+ */
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply controller/sharding/sharding.go:194]: The multiline comments do work in Go. But I can surely update it to use `//`

File: controller/sharding/sharding.go [hunk 14]
```diff
@@ -162,3 +186,156 @@ func createClusterIndexByClusterIdMap(db db.ArgoDB) map[string]int {
...
+func GetShardFromConfigMap(kubeClient *kubernetes.Clientset, settingsMgr *settings.SettingsManager, replicas, shard int) (int, error) {
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply controller/sharding/sharding.go:195]: The goal of the function is to basically just to compute and get the shard from the ConfigMap. `GetOrUpdateShardFromConfigMap` sounds good too.

File: controller/sharding/sharding.go (hunk 3 above)
Comment (ishitasequeira) [reply controller/sharding/sharding.go:209]: So, the users do have a capability to set the shard number for a specific controller today. The function takes the shard number from the environment variable (default value -1, if not set) and passes it to this function. So, if the shard value passed to this function is `-1`, that is, the shard was not set as an environment variable, we default the shard number to 0 for computing the default config map.

File: controller/sharding/sharding.go [hunk 15]
```diff
@@ -162,3 +186,156 @@ func createClusterIndexByClusterIdMap(db db.ArgoDB) map[string]int {
...
+		}
+
+		shard, shardMappingData := getOrUpdateShardNumberForController(shardMappingData, hostname, replicas, shard)
+		updatedShardMappingData, err := json.Marshal(shardMappingData)
+		if err != nil {
+			return -1, fmt.Errorf("error marshalling data of shard mapping ConfigMap: %s", err)
+		}
+		shardMappingCM.Data[ShardControllerMappingKey] = string(updatedShardMappingData)
+
+		_, err = kubeClient.CoreV1().ConfigMaps(settingsMgr.GetNamespace()).Update(context.Background(), shardMappingCM, metav1.UpdateOptions{})
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply controller/sharding/sharding.go:237]: If we face an error here, especially the conflict error, we need to re-compute the shard number again from the ConfigMap. So, I have added the retry on the top of the entire logic [here](https://github.com/argoproj/argo-cd/blob/58c04cb4b09a4bd66093b2be1e3608dfe52ee301/cmd/argocd-application-controller/commands/argocd_application_controller.go#L230-L242).

File: manifests/base/application-controller/argocd-application-controller-deployment.yaml [hunk 16]
```diff
@@ -1,5 +1,5 @@
 apiVersion: apps/v1
-kind: StatefulSet
+kind: Deployment
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply manifests/base/application-controller/argocd-application-controller-deployment.yaml:2]: By `current sharding balance logic` do you mean the cluster distribution or the mechanism to compute the shard for each controller?

File: util/db/db.go [hunk 17]
```diff
@@ -83,6 +87,9 @@ type ArgoDB interface {
 	AddGPGPublicKey(ctx context.Context, keyData string) (map[string]*appv1.GnuPGPublicKey, []string, error)
 	// DeleteGPGPublicKey removes a GPG public key from the configuration
 	DeleteGPGPublicKey(ctx context.Context, keyID string) error
+
+	// GetApplicationControllerReplicas gets the replicas of application controller
+	GetApplicationControllerReplicas() int
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply util/db/db.go:92]: That sounds good. I will update it.

File: util/db/db.go [hunk 18]
```diff
@@ -140,3 +147,14 @@ func (db *db) unmarshalFromSecretsStr(secrets map[*SecretMaperValidation]*v1.Sec
 func StripCRLFCharacter(input string) string {
 	return strings.TrimSpace(input)
 }
+
+// GetApplicationControllerReplicas gets the replicas of application controller
+func (db *db) GetApplicationControllerReplicas() int {
+	// get the replicas from application controller deployment, if the application controller deployment does not exist, check for environment variable
+	applicationControllerName := env.StringFromEnv(common.EnvAppControllerName, common.DefaultApplicationControllerName)
+	appControllerDeployment, _ := db.kubeclientset.AppsV1().Deployments(db.settingsMgr.GetNamespace()).Get(context.Background(), applicationControllerName, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply util/db/db.go:155]: As I mentioned in other comment, we do not have a Deployment Informer today. I can check if I can add one. 

I think we would not want to cache the deployments but instead get the latest deployment always to make sure we fetch the correct replica count.

**Review by ishitasequeira — COMMENTED** [review]
Thanks for the detailed review @leoluz !!

I am working on the documentation and adding the quality unit tests to the PR.

File: controller/sharding/sharding.go [hunk 19]
```diff
@@ -162,3 +186,156 @@ func createClusterIndexByClusterIdMap(db db.ArgoDB) map[string]int {
...
+		if err != nil {
+			return -1, err
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply controller/sharding/sharding.go:239]: Didn't add context here as we need to check for the Conflict error at the place of the caller of the function where we have the retry logic. I have added the context there.

File: cmd/main.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply cmd/main.go:38]: Actually, I will keep this a string and maybe create a separate PR for changing this to constants.

File: util/db/db.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (leoluz) [reply util/db/db.go:155]: The informer will return you the latest state as well with the advantage to avoid making an additional call to KubeAPI

**Comment by ishitasequeira** [comment]
@leoluz addressed your comments. PTAL.

File: docs/operator-manual/high_availability.md [hunk 20]
```diff
@@ -60,15 +60,14 @@ reconciliation. In this case, we advise to use the preferred resource version in
 * The controller polls Git every 3m by default. You can change this duration using the `timeout.reconciliation` setting in the `argocd-cm` ConfigMap. The value of `timeout.reconciliation` is a duration string e.g `60s`, `1m`, `1h` or `1d`.
 
 * If the controller is managing too many clusters and uses too much memory then you can shard clusters across multiple
-controller replicas. To enable sharding increase the number of replicas in `argocd-application-controller` `StatefulSet`
-and repeat the number of replicas in the `ARGOCD_CONTROLLER_REPLICAS` environment variable. The strategic merge patch below
+controller replicas. To enable sharding increase the number of replicas in `argocd-application-controller` `Deployment`. The strategic merge patch below
```
Comment (leoluz) [inline docs/operator-manual/high_availability.md:63]: I would be great to have something in the docs describing how the dynamic shard rebalance works. Things like:
- How to enable it?
- What is the expected behaviour?
- What are the different methods of rebalancing shards implemented in Argo CD today?
- What is the guideline to switch the controller from StatefulSet to Deployment?

File: cmd/argocd-application-controller/commands/argocd_application_controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (leoluz) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go:222]: Thank you. Was this added? I didn't find it in the code.

**Review by leoluz — CHANGES_REQUESTED** [review]

File: controller/sharding/sharding_test.go [hunk 21]
```diff
@@ -430,3 +431,251 @@ func Test_generateDefaultShardMappingCM_PredefinedShard(t *testing.T) {
...
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			defer func() { osHostnameFunction = os.Hostname }()
+			heartbeatCurrentTime = func() metav1.Time { return expectedTime }
+			shard, shardMappingData := getOrUpdateShardNumberForController(tc.shardApplicationControllerMapping, tc.hostname, tc.replicas, tc.shard)
+			assert.Equal(t, tc.expectedShard, shard)
+			// for i, shardMapping := range shardMappingData {
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply controller/sharding/sharding_test.go:675]: I am validating the shardMapping using the line 
```
assert.Equal(t, tc.expectedShardMappingData, shardMappingData)
```

Missed removing the commented code.

File: controller/appcontroller.go [hunk 22]
```diff
@@ -204,9 +208,27 @@ func NewApplicationController(
...
+	// 			return
+	// 		}
+	// 		oldDepl, oldOK := old.(*apiv1.Deployment)
+	// 		newDepl, newOK := new.(*apiv1.Deployment)
+	// 		if oldOK && newOK && oldDepl.Spec.Replicas != nil && newDepl.Spec.Replicas != nil && oldDepl.Spec.Replicas != newDepl.Spec.Replicas {
+	// 			// Update Cache
+	// 			log.Debugf("Reached here")
+	// 		}
+	// 	},
+	// })
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply controller/appcontroller.go]: This commented code is already removed.

File: cmd/argocd-application-controller/commands/argocd_application_controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go:222]: Yes.. This was added to code [here](https://github.com/argoproj/argo-cd/pull/15036/files#diff-21d1e1ae7561765d523f51ae68a0b94580d1a5980378aa8ccbaa2a7581fc0643R221)

File: controller/sharding/sharding.go (hunk 3 above)
Comment (ishitasequeira) [reply controller/sharding/sharding.go:213]: Done

File: docs/operator-manual/high_availability.md (hunk 20 above)
Comment (ishitasequeira) [reply docs/operator-manual/high_availability.md:63]: > * How to enable it?
It would be enabled by default when users upgrade to v2.9, as it will deploy application controller as a deployment instead of a statefulset.

> What is the expected behaviour?
I added some detail around it.

> What are the different methods of rebalancing shards implemented in Argo CD today?
I think this would be not needed as we already have sharding documented under the argocd-application-controller section [here](https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#argocd-application-controller)

> What is the guideline to switch the controller from StatefulSet to Deployment?
IMO, it would be more useful to have this as part of upgrade notes instead of having it here. WDYT?

File: cmd/argocd-application-controller/commands/argocd_application_controller.go [hunk 23]
```diff
@@ -208,20 +211,49 @@ func NewCommand() *cobra.Command {
 }
 
 func getClusterFilter(kubeClient *kubernetes.Clientset, settingsMgr *settings.SettingsManager, shardingAlgorithm string) sharding.ClusterFilterFunction {
-	replicas := env.ParseNumFromEnv(common.EnvControllerReplicas, 0, 0, math.MaxInt32)
+
+	var replicas int
 	shard := env.ParseNumFromEnv(common.EnvControllerShard, -1, -math.MaxInt32, math.MaxInt32)
+
+	applicationControllerName := env.StringFromEnv(common.EnvAppControllerName, common.DefaultApplicationControllerName)
+	appControllerDeployment, _ := kubeClient.AppsV1().Deployments(settingsMgr.GetNamespace()).Get(context.Background(), applicationControllerName, metav1.GetOptions{})
```
Comment (crenshaw-dev) [inline cmd/argocd-application-controller/commands/argocd_application_controller.go:219]: Looks like there's an unhandled error here?

Comment (ishitasequeira) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go:219]: Actually I didn't add a check on the error here as we only use the new logic if we found an application-controller Deployment. If we do not find a deployment for whatever reason, we use the existing logic of using Statefulset.

File: manifests/base/application-controller/argocd-application-controller-deployment.yaml (hunk 16 above)
[the remark this answers is not in the document]
Comment (leoluz) [reply manifests/base/application-controller/argocd-application-controller-deployment.yaml:2]: The distribution algorithm. The default one depends on the statefulset to work irrc.

File: docs/operator-manual/dynamic-cluster-distribution.md [hunk 24]
```diff
@@ -0,0 +1,283 @@
...
+Starting v2.9, ArgoCD supports a dynamic cluster distribution feature. In this mechanism, Argo CD sharding uses Deployments for the application controller. 
+
+
+## Enabling Dynamic Distribution of Clusters
+
+Inorder to utilize the feature, the StatefulSet of Application Controller needs to be replaced with the Deployment Configuration of Application Controller. To do so, set the number of replicas of StatefulSet and the environment variable `ARGOCD_CONTROLLER_REPLICAS` to 0 disabling all current application controllers.
+
+Once all the current controllers have stopped working, apply the below deployment configuration of application controller with the desired number of replicas.
+
+```yaml
```
Comment (leoluz) [inline docs/operator-manual/dynamic-cluster-distribution.md:21]: I think it is better to create a dedicated overlay in the `manifests` folder to provide this new deployment. Later on we could promote this overlay to the base. WDYT?

File: docs/operator-manual/dynamic-cluster-distribution.md [hunk 25]
```diff
@@ -0,0 +1,283 @@
...
+          - key: ca.crt
+            path: ca.crt
+```
+
+Note the introduction of new environment variable `ARGOCD_CONTROLLER_HEARTBEAT_TIME`. The environment variable is explained in [working of Dynamic Distribution Heartbeat Process](#working-of-dynamic-distribution)
+
+
+## Working of Dynamic Distribution
+
+Along with the new mechanism of sharding using Deployments for Application Controller, Application Controller also came up with new format for managing Controller <-> Shard mappings.
```
Comment (leoluz) [inline docs/operator-manual/dynamic-cluster-distribution.md:268]: Maybe rephrase to something like:

Along with the new sharding strategy using Deployments for the Application Controller there is a new mechanism to dynamically associate clusters to specific Controller replicas in the shard.

File: docs/operator-manual/dynamic-cluster-distribution.md [hunk 26]
```diff
@@ -0,0 +1,283 @@
...
+ControllerName : "argocd-application-controller-hydrxyt"
+HeartbeatTime  : "2009-11-17 20:34:58.651387237 +0000 UTC"
+```
+
+* `ControllerName`: Stores the hostname of the Application Controller pod
+* `ShardNumber` : Stores the shard number managed by the controller pod
+* `HeartbeatTime`: Stores the last time this heartbeat was updated.
+
+
+Controller Shard Mapping is updated in the configMap during each readiness probe check of the pod, that is every 10 seconds (otherwise as configured). The controller will acquire the pod during every iteration of readiness probe check and try to update the ConfigMap with the `HeartbeatTime`. The default `HeartbeatDuration` after which the heartbeat should be updated is `10` seconds. If the ConfigMap was not updated for any controller pod for more than `3 * HeartbeatDuration`, then the readiness probe for the application pod is marked as `Unhealthy`. To increase the default `HeartbeatDuration`, you can set the environment variable `ARGOCD_CONTROLLER_HEARTBEAT_TIME` with the desired value.
```
Comment (leoluz) [inline docs/operator-manual/dynamic-cluster-distribution.md:283]: Not sure what "The controller will acquire the pod" means exactly. Can you please clarify?

File: docs/operator-manual/dynamic-cluster-distribution.md [hunk 27]
```diff
@@ -0,0 +1,283 @@
...
+          - key: tls.key
+            path: tls.key
+          - key: ca.crt
+            path: ca.crt
+```
+
+Note the introduction of new environment variable `ARGOCD_CONTROLLER_HEARTBEAT_TIME`. The environment variable is explained in [working of Dynamic Distribution Heartbeat Process](#working-of-dynamic-distribution)
+
+
+## Working of Dynamic Distribution
```
Comment (leoluz) [inline docs/operator-manual/dynamic-cluster-distribution.md:266]: I think it is important to mention how adding new clusters to Argo CD and increasing the number of controller replicas will trigger the rebalancing logic without requiring any manual procedure like before.

**Review by leoluz — CHANGES_REQUESTED** [review]
Docs are looking awesome. Thank you Ishita!
Added few final comments in the docs.

File: docs/operator-manual/dynamic-cluster-distribution.md (hunk 26 above)
Comment (ishitasequeira) [reply docs/operator-manual/dynamic-cluster-distribution.md:283]: Ahh.. it is shard. Missed it out.. Updating.

File: docs/operator-manual/dynamic-cluster-distribution.md (hunk 24 above)
Comment (ishitasequeira) [reply docs/operator-manual/dynamic-cluster-distribution.md:21]: Good idea.

File: docs/operator-manual/dynamic-cluster-distribution.md [hunk 28]
```diff
@@ -6,266 +6,24 @@ But using StatefulSet has the following limitations:
...
 
 
 Starting v2.9, ArgoCD supports a dynamic cluster distribution feature. In this mechanism, Argo CD sharding uses Deployments for the application controller. 
 
 
 ## Enabling Dynamic Distribution of Clusters
 
-Inorder to utilize the feature, the StatefulSet of Application Controller needs to be replaced with the Deployment Configuration of Application Controller. To do so, set the number of replicas of StatefulSet and the environment variable `ARGOCD_CONTROLLER_REPLICAS` to 0 disabling all current application controllers.
+Inorder to utilize the feature, the StatefulSet of Application Controller needs to be replaced with the Deployment Configuration of Application Controller. To do so, set the number of replicas of StatefulSet and the environment variable `ARGOCD_CONTROLLER_REPLICAS` to 0 disabling all current application controllers and deploy application controller as a deployment.
```
Comment (leoluz) [inline docs/operator-manual/dynamic-cluster-distribution.md:17]: typo: `s/Inorder/In order`

File: docs/operator-manual/dynamic-cluster-distribution.md (hunk 27 above)
Comment (leoluz) [reply docs/operator-manual/dynamic-cluster-distribution.md:24]: @ishitasequeira If adding new clusters to Argo CD doesn't trigger the rebalancing logic how it will be associated to one of the shards?

**Review by leoluz — COMMENTED** [review]
Found one small typo in the docs and asked one more question.

File: docs/operator-manual/dynamic-cluster-distribution.md (hunk 27 above)
Comment (ishitasequeira) [reply docs/operator-manual/dynamic-cluster-distribution.md:24]: So, adding new clusters will follow the current legacy or round-robin algorithm whichever is defined in the controller.

File: manifests/base/application-controller/argocd-application-controller-deployment.yaml (hunk 16 above)
[the remark this answers is not in the document]
Comment (leoluz) [reply manifests/base/application-controller/argocd-application-controller-deployment.yaml]: Ok. After talking with @ishitasequeira it was clarified that the current distribution algorithm will work with deployment as it will also rely on the new configmap mechanism.

**Review by leoluz — APPROVED** [review]
LGTM. Thank you for this feature Ishita!

Note: There is still an open discussion to whether or not moving the Argo CD Controller to `Deployment` is the direction we should go. It would cause unnecessary burden in our user base by asking all Argo CD users to do migrations if not justified. If the same feature can be implemented keeping the current controller as `StatefulSet` it would avoid unnecessary migrations unless there are good and solid reasons to move to `Deployment`.
